group/assignment5/WordDependencyGraph.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <iostream>
+#include <vector>
+#include <unordered_map>
+#include <set>
+
+using namespace std;
+
+class WordDependencyGraph{
+	unordered_map<char, vector<char>> graph;
+public:
+	WordDependencyGraph(vector<string>& dictionary) {
+		generateGraph(dictionary);
+	}
+	
+	vector<char> topologicalSort(){
+		vector<char> lexicographicalOrder;
+
+		set<char> visited;
+		
+		for (unordered_map<char, vector<char>>::iterator it = graph.begin(); it != graph.end(); it++) {
+			char currentChar = it->first;
+			if (visited.find(currentChar)==visited.end())
+				topologicalSortUtil(currentChar, visited, lexicographicalOrder);
+		}
+
+		return reverseOrder(lexicographicalOrder);
+	}
+
+private:
+
+	void topologicalSortUtil(char key, set<char>& visited, vector<char>& lexicographicalOrder) {
+		visited.insert(key);
+		vector<char> children = graph[key];
+
+		for (char ch : children) {
+			if (visited.find(ch) == visited.end())
+				topologicalSortUtil(ch, visited, lexicographicalOrder);
+		}
+
+		lexicographicalOrder.push_back(key);
+	}
+
+	vector<char> reverseOrder(vector<char>& lexicographicalOrder) {
+		vector<char> lexicographicalOrderReveresed;
+
+		for (int i = lexicographicalOrder.size() - 1; i >= 0; i--)
+			lexicographicalOrderReveresed.push_back(lexicographicalOrder[i]);
+
+		return lexicographicalOrderReveresed;
+	}
+
+	void generateGraph(vector<string>& dictionary) {
+		int maxWordLength = getMaxLength(dictionary);
+
+		for (int s = 0; s < maxWordLength; s++) {
+			for (int i = 0; i < dictionary.size(); i++) {
+				string word1 = dictionary[i];
+				for (int j = i + 1; j < dictionary.size(); j++) {
+					string word2 = dictionary[j];
+
+					if (word1.length() > s && word2.length() > s && haveTheSamePrefix(word1, word2, s)) {
+						char ch1 = word1[s];
+						char ch2 = word2[s];
+
+						if (ch1 != ch2) {
+							// ch1 should be ordered before ch2 => ch2 is dependent on ch1
+							// insert ch2 as a child of ch1 node if it is not already inserted
+							vector<char>& childrenOfChar1 = graph[ch1];
+
+							if (childrenOfChar1.size() > 0) {
+								// check if ch2 is already inserted as a child of ch1
+								if (!containsChar(childrenOfChar1, ch2)) {
+									childrenOfChar1.push_back(ch2);
+								}
+							}
+							else {
+								childrenOfChar1.push_back(ch2);
+							}
+						}
+					}
+				}
+			}
+		}
+		// iteration for the last word in the dictionary
+		string word1 = dictionary[dictionary.size()-1];
+		for (char ch : word1) {
+			// if graph does not contain pair with the first element as letter ch, letter ch will be inserted with an empty children list,
+			// by just referencing it.
```

Comment:
I think you need to do this not only for the last word, but make sure to insert into your graph all characters that aren't prior to others.
Your test works because the only characters that are not oredered before any other character are the ones in the last word, so you eventually find all the letters.
For this dictionary: {"abd", "abc", "^pap", "^alk", "*$llo", "car", "01o"}
the algorithm doesn't put all of the characters inside the alphabet, since, for example, '$' isn't prior to any other character and also isn't part of the last word, so it doesn't get into the alphabet.

Reply:
Yes, you are right :) Thank you! I will correct this.